### PR TITLE
dyno: initial steps for producing typed AST from uAST + types

### PR DESCRIPTION
**Pull request by mppf**
This PR adds a new prototype for converting from uAST + types to production compiler AST. The dyno effort to improve the compiler has been recently focusing on creating an incremental resolver. We would like to use this incremental resolver in production `chpl` compiles, but we will need to convert its results into a form that the production compiler understands to do so.

This PR also improves the user-facing error message for mixing `return;` and `return x;` in one function.

Before this PR, `convert-uast.cpp` had some initial support for converting types and resolved calls. However, it was hard to see how to continue to extend this in-place because both the input and output of this compiler pass need to change:
 * the input for `convert-uast.cpp` is an untyped uAST traversal but it needs to be a typed uAST traversal using ResolvedVisitor
 * the output from `convert-uast.cpp` is predominantly untyped AST but it needs to be typed AST with other changes.

This PR changes direction by adding `convert-typed-uast.cpp`. This is formed in part by copying some of the code from `convert-uast.cpp` and modifying it. Besides having a tighter implementation, there are two key ideas here:
 1. We will need to be able to make incremental progress with the implementation. To enable that, this PR computes a call graph of functions called from user code. Only the functions in this call graph will go through the new strategy.
 2. The AST produced by `convert-typed-uast.cpp` will match the compiler's AST after `callDestructors`. That means that `convert-typed-uast.cpp` will work with the dyno resolver to normalize and to handle copy-init and deinit calls.

Implementation Details:
 * Some code can be reasonably shared between `convert-uast.cpp` and `convert-typed-uast.cpp`. I moved that code to `convert-help.cpp`.
 * Since there are now two uAST -> AST converters that need to potentially interact or to replace each other, these are now subclasses of an abstract base class.
 * Since `convert-uast.cpp` no longer needs to handle types, I removed the type handling code from there (and much of it is added back in a slightly different form in `convert-typed-uast.cpp`).
 * To allow simple programs to compile end-to-end (and enable incremental improvements to get more and more of them working), this PR adds a call-graph analysis and uses the call graph of function called from user module code in order to decide which functions to convert in this new manner when compiling with `--dyno`.

Future Work:
 * handle `proc main` (both converting it if present & also creating the compiler-generated one)
 * adjust `call-graph.cpp` and `convert-typed-uast.cpp` to handle nested functions
 * fill in `convert-typed-uast.cpp` for many code patterns (this PR is a starting point, and calls and module init functions seem to be working OK, but most uAST patterns are unimplemented). 
 * fill in the implementation for `operator <` for `QualifiedType` to check the contents of `Type*`s.
 * adjust later passes to ignore processing functions with already-resolved AST as generated by `convert-typed-uast.cpp`.

Reviewed by @DanilaFe - thanks!

- [x] full comm=none testing

### CODE REVIEW

File: frontend/lib/resolution/call-graph.cpp
Comment (DanilaFe) [inline frontend/lib/resolution/call-graph.cpp:123]: Am I right to assume that this doesn't consider nested functions? It is my understanding that the resolutioncontext is supposed to keep track of the "call stack" of functions that were used to reach a given fn. But here, we're constructing a new resolution context from scratch, which means that called nested functions will lose their nested function calling context.

File: frontend/lib/resolution/resolution-types.cpp
Comment (DanilaFe) [inline frontend/lib/resolution/resolution-types.cpp:74]: Could you explain why all these `<`s are needed?

File: frontend/lib/uast/post-parse-checks.cpp
Comment (DanilaFe) [inline frontend/lib/uast/post-parse-checks.cpp:1978]: This is called `CountReturnsYields` but only seems to count returns. Moreover, it's certainly not equivalent to the yield / return checks that were removed from the production AST checking code. Did you mean to include that in this PR as well? Alternatively, is there a reason the checks are no longer needed?

File: compiler/passes/convert-typed-uast.cpp
Comment (DanilaFe) [inline compiler/passes/convert-typed-uast.cpp:90]: This is duplicated verbatim from `convert-uast.cpp`. Consider making it an associated / inner class of `UastConverter` or just lifting it to a top-level struct.

File: compiler/passes/convert-typed-uast.cpp
Comment (DanilaFe) [inline compiler/passes/convert-typed-uast.cpp:136]: `rvv` is the return value temp, right? I think moving away from the name `rvv` would be helpful here.

**Review by DanilaFe — COMMENTED** [review]
I reviewed everything but the "meat" (`convert-typed-uast.cpp`). I expect that to take me some time, so here are my comments in the meantime.

File: compiler/passes/convert-typed-uast.cpp
Comment (DanilaFe) [inline compiler/passes/convert-typed-uast.cpp:725]: Here too I worry about missing resolution state from the parent function.

File: compiler/passes/convert-typed-uast.cpp
Comment (DanilaFe) [inline compiler/passes/convert-typed-uast.cpp:1505]: I think this is broken. My reasoning: above, you rely on `symbol`, which is a global variable. `convertModuleInit` performs "cleanup" after it's done, resetting global variables, including resetting `symbol` to `nullptr`. As such, I think it clobbers important state. That makes me think you need to preserve that kind of state and restore it (instead of clobbering).

**Review by DanilaFe — COMMENTED** [review]
Having gotten through through `convert-type-uast.cpp` my chief concern is the amount of duplication between the typed and untyped conversions. I know that code was copied (I see David's comments that I remember from `convert-uast.cpp`, but it also looks significantly changed. I find this concerning because, given the complexity of typed and untyped conversions, it's relatively hard for me to observe any differences between the two.

This is true in general. I can read through these `enter` functions (I have given them only brief looks while reviewing), but I certainly can't tell you to what degree they are equivalent to their untyped versions. And if I were to fix a bug or make a change in the sibling pass (e.g., typed to untyped), I think I would have some difficulty ensuring that change occurs in the current pass.

I don't know what can be done to address these changes -- can we split up some complicated transformations into structs/classes with state, and implement the calls to `enter` in terms of method calls on such structs, which incrementally change state? We might then be able to interweave typed-specific logic with the general conversion shape. I am okay with the PR going in as-is in the name of progress, but I do think it's going to be a pretty significant maintenance challenge.

File: compiler/include/convert-uast.h
Comment (mppf) [reply compiler/include/convert-uast.h:35]: We repeat it for a lot of the methods e.g. `TConverter::enter(...)` and so wanted to use a short name to keep all of the formals from being too far to the right.

File: frontend/lib/resolution/call-graph.cpp
Comment (mppf) [reply frontend/lib/resolution/call-graph.cpp:123]: Yes. I developed this code before David's recent work was merged. We can leave adjusting it to handle nested functions correctly as future work. It might require some changes more on the nested functions infrastructure side.

File: frontend/lib/resolution/resolution-types.cpp
Comment (mppf) [reply frontend/lib/resolution/resolution-types.cpp:74]: It's for `TConverter::convertFunctionsToConvert`. That function converts all functions in a `CalledFnsSet` (`CalledFnsSet` is a set of functions = `std::unordered_map<const ResolvedFunction*, int>`). Because production AST does things like encode the order of declaration in the integer `id`; and because follow-on passes can depend on the ordering of AST declarations; we need this function to convert functions in the same order on every compilation, even though the `unordered_map` iteration order would be nondeterministic.

File: frontend/lib/uast/post-parse-checks.cpp
Comment (mppf) [reply frontend/lib/uast/post-parse-checks.cpp:1978]: I can rename it to `CountReturns`.

Those checks already existed in the form of checks within `checkParentsForControlFlow`. In fact I had at first assumed that `checkParentsForControlFlow` covered all of them, but the `Not all returns in this function return a value` error was not (as revealed by testing).

Are the others covered? I can check:
 * `return statement is not in a function or iterator` -- covered by `'return' is not allowed outside of a procedure or an iterator`
 * `returning a value in an iterator` -- covered by `return' statements with values are not allowed in iterators`
 * `yield statement is outside an iterator` -- covered by `'yield' is not allowed outside of an iterator`
 * `yield statement is in a non-iterator function` -- covered by `'yield' is not allowed outside of an iterator`

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp:90]: I will add a TODO comment. The issue is that the typed converter doesn't yet handle variable multi-decls, and while it probably needs something kindof like this, it's very unlikely to be exactly this. In particular, it should be using some sort of `Type*` / `QualifiedType` rather than `Expr* prevTypeExpr`.

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp:136]: `rvv` is specific and used in the production compiler. Most concretely, the return value variable is the variable with `FLAG_RVV`.

I will add a comment to say that it's the return value variable.

File: frontend/lib/resolution/call-graph.cpp
Comment (mppf) [reply frontend/lib/resolution/call-graph.cpp:123]: I've added this to a Future Work section of the PR message.

**Comment by mppf** [comment]
> Having gotten through through `convert-type-uast.cpp` my chief concern is the amount of duplication between the typed and untyped conversions. I know that code was copied (I see David's comments that I remember from `convert-uast.cpp`, but it also looks significantly changed. I find this concerning because, given the complexity of typed and untyped conversions, it's relatively hard for me to observe any differences between the two.
> 
> This is true in general. I can read through these `enter` functions (I have given them only brief looks while reviewing), but I certainly can't tell you to what degree they are equivalent to their untyped versions. And if I were to fix a bug or make a change in the sibling pass (e.g., typed to untyped), I think I would have some difficulty ensuring that change occurs in the current pass.
> 
> I don't know what can be done to address these changes -- can we split up some complicated transformations into structs/classes with state, and implement the calls to `enter` in terms of method calls on such structs, which incrementally change state? We might then be able to interweave typed-specific logic with the general conversion shape. I am okay with the PR going in as-is in the name of progress, but I do think it's going to be a pretty significant maintenance challenge.

I struggled with adjusting `convert-uast.cpp` for a long time. I wanted to make incremental progress on it but kept getting caught up in problems with the structure of that pass. To me, splitting the typed conversion into a different file is what allows any progress in this area at all. When it comes down to it, they are doing very different things and they have different issues.

>  I can read through these `enter` functions (I have given them only brief looks while reviewing), but I certainly can't tell you to what degree they are equivalent to their untyped versions.

Of course they are different. The new ones are doing typed conversion!

> And if I were to fix a bug or make a change in the sibling pass (e.g., typed to untyped), I think I would have some difficulty ensuring that change occurs in the current pass.

IMO the form of the bug fix would be very different between the two, because having type information available everywhere is a very different situation. Moreover, the `AST` that is being generated is totally different. `convert-typed-uast.cpp` generates *normalized and resolved AST* while `convert-uast.cpp` matches the AST from ye olde parser while adding in some scope resolution. As a specific example, `convert-typed-uast.cpp` will handle `return 1` by storing `1` into the return value variable, rather than emitting a `PRIM_RETURN` and expecting the normalizer to change it later.

I would even go so far as to argue that a given bug is unlikely to affect both of them.

Another element here is that I don't expect that we'll keep `convert-uast.cpp` at all in the long term. Once the new resolver is complete and we have completed `convert-typed-uast.cpp`, we can remove `convert-uast.cpp` and put any bits of it that are still used by `convert-typed-uast.cpp` into that remaining file. It has always been our plan to replace the early passes in the compiler & IMO the strategy here with `convert-typed-uast.cpp` will help us to do that.

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp:547]: IMO supporting `<` is not conceptually complicated. It is just that it impacts many types. Another thought about it is that, it seems somewhat likely to me that we will run into other cases where `<` on some of these types will be needed. One specific case that I can think of: I think we are going to need to sort the instantiations when creating a separate compilation library file (for stable output). But, if we have them sorted, and we have `<` on them, then we can do binary search when looking for a particular instantiation.

Now, I did not implement `<` between types. (I left that as a TODO). How exactly to do that seems somewhere between tricky and annoying to implement, so I am sympathetic to your point.

> can you just sort by ID, and then keep insertion order? Since insertion order is determined by a call graph traversal, I'd expect that to be stable enough

It may be already obvious to you already, but the reason that sorting by ID isn't good enough is that we have to concern ourselves with instantiations of generic functions. It might be possible to track the order in which these were instantiated. We couldn't do that within the types themselves, because that gets into issues with the query framework (on a recompilation, I'd want it to reuse the same instantiation even if it went from being instantiation 1 to instantiation 3).

However it is an interesting idea to include it in the call graph traversal. Were I to do that, would you favor keeping the `<` that I implemented (in case they are useful later) or removing them (so as to avoid any maintenance burden)?

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp:1507]: I think you are right that there is an issue here in order to handle submodules correctly. However I know that doesn't work (and neither do many other patterns).

In an earlier version, I was saving this state, but I convinced myself that it's better if the traversal just visits one function / module at a time. That works OK for functions and modules with the code I have here, but not for submodules. To handle those in this way, we can accumulate the submodules to visit in a list and then process them after visiting the enclosing module. I can look at doing that here.

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp:727]: Added a note to the PR's future work section.

File: compiler/passes/convert-typed-uast.cpp
Comment (DanilaFe) [reply compiler/passes/convert-typed-uast.cpp]: I would certainly favor removing them (if you needed them later, you could always find the diff in this PR), but I will not block the PR over their inclusion. If you think they are needed after all (or will be needed in the future), then I trust your judgement.

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp]: On second thought, I'll rename it to `curRetVar`, since it is a field, and `rvv` might be too obtuse when looking at methods here.

File: compiler/passes/convert-typed-uast.cpp
Comment (mppf) [reply compiler/passes/convert-typed-uast.cpp:430]: I opted to leave these as-is for now.

**Review by DanilaFe — APPROVED** [review]

**Pull request merged**
